src/Renderer/SwiperRenderer.php

```diff
@@ -0,0 +1,205 @@
+        }
+
+        if ($model->sliderSpaceBetween) {
+            $params['spaceBetween'] = (int) $model->sliderSpaceBetween;
+        }
+
+        if ('crossfade' === $model->sliderEffect) {
+            $params['effect'] = 'fade';
+            $params['fadeEffect'] = ['crossFade' => true];
+        } elseif ($model->sliderEffect) {
+            $params['effect'] = $model->sliderEffect;
+        }
+
+        if ($model->sliderContinuous) {
+            $params['loop'] = true;
+        }
+
+        if ($model->sliderButtons) {
+            $params['navigation'] = [
+                'nextEl' => '#'.$swiperId.' .swiper-button-next',
+                'prevEl' => '#'.$swiperId.' .swiper-button-prev',
+            ];
+        }
+
+        if ($model->sliderPagination) {
+            $params['pagination'] = [
+                'el' => '#'.$swiperId.' .swiper-pagination',
+                'clickable' => true,
+            ];
+
+            if ($model->sliderPaginationType) {
+                // backwards compatibility
+                $model->sliderPaginationType = 'progress' === $model->sliderPaginationType ? 'progressbar' : $model->sliderPaginationType;
+                $params['pagination']['type'] = $model->sliderPaginationType;
+            }
+        }
+
+        if ($model->sliderScrollbar) {
+            $params['scrollbar'] = [
+                'el' => '#'.$swiperId.' .swiper-scrollbar',
+                'draggable' => true,
+            ];
+        }
+
+        $params = $this->addBreakpointsToParams($model, $params);
+
+        if ($model->sliderAutoheight) {
+            $params['autoHeight'] = true;
+        }
+
+        if ($model->sliderCenteredSlides) {
+            $params['centeredSlides'] = true;
+        }
+
+        if ($model->sliderCustomOptions && null !== ($customOptions = json_decode($model->sliderCustomOptions, true))) {
+            $params = array_replace_recursive($params, $customOptions);
+        }
+
+        $template->sliderButtons = $model->sliderButtons;
+        $template->sliderPagination = $model->sliderPagination;
+        $template->sliderScrollbar = $model->sliderScrollbar;
+        $template->wrapperClass = $model->sliderWrapperClass;
+        $template->parameters = $params;
+        $template->sliderId = $swiperId; // unique name for an entry in the sliderConfig-variable
+    }
+
+    public function addAssets(): void
+    {
+        // check if the scripts should be combined
+        $combine = '';
+
+        // get the current page
+        $page = null;
+        $request = $this->requestStack->getCurrentRequest();
+
+        if ($request) {
+            /** @var PageModel $page */
+            $page = $request->attributes->get('pageModel');
+        }
+        // if there is no request or "pageModel" is not part of the request-attributes
+        // use the $GLOBALS['objPage']
+        if (!$page instanceof PageModel) {
```

Comment:
@fritzmg this validation is necessary for Contao 4.13 because in my testsystem the pageModel of the request was a string! I'm not sure if it's a bug in the core 🤔

Reply:
The `pageModel` request attribute can be a number (the ID of the model) in fragments. So if you used the service from within a fragment that would explain it.

Reply:
yep, we have a custom fragment module which uses the `SwiperRenderer` ☺️